### PR TITLE
Add Noms commit command

### PR DESCRIPTION
**Pull request by mikegray**
towards #2100

### CODE REVIEW

**Comment by mikegray** [comment]
@aboodman please provide feedback as you have time, continuing to add tests and command line arguments

File: cmd/noms/noms_commit.go [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+		commitFlagSet.StringVar(&metaP, "meta-p", "", "'<key>:<path>' - creates a metadata field called 'key' set to the value at <path>")
+	*/
+	commitFlagSet.StringVar(&message, "message", "", "alias for -meta 'message:<message>'")
+	commitFlagSet.StringVar(&date, "date", "", "alias for -meta 'date:<date>'. '<date>' must be iso8601-formatted. If '<date>' is empty, it defaults to the current date.")
+	return commitFlagSet
+}
+
+func runCommit(args []string) int {
+	ds, err := spec.GetDataset(args[0])
+	d.CheckError(err)
```
Comment (aboodman) [inline cmd/noms/noms_commit.go]: Make this a more friendly error.

Comment (aboodman) [inline cmd/noms/noms_commit.go]: Oh nevermind, I this was d.Chk.

File: cmd/noms/noms_commit.go [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+	if pathResolvedValue == nil {
+		d.CheckErrorNoUsage(fmt.Errorf("Object not found: %s", args[1]))
+	}
+
+	// Note that spec.GetPath() above gave us pathResolvedValue by calling
+	// ReadValue() already which loaded the cache for value into pathDb,
+	// but we are going to commit to ds.Database() so we need the cache for
+	// pathResolvedValue loaded into ds.Database() to perform the commit
+	// successfully
+	value := ds.Database().ReadValue(pathResolvedValue.Hash())
```
Comment (aboodman) [inline cmd/noms/noms_commit.go]: This won't work in the general case (`pathResolvedValue.Hash()`) is not guaranteed to exist in target database.

I think you want to arrange things such that you load the path and commit to the same database instance. Probably the way to do this is to make the path argument a relative, not absolute. Then load the dataset the normal way, then parse the relative path, then resolve it against the dataset root.

Comment (aboodman) [inline cmd/noms/noms_commit.go]: Then in the code, you'd use `spec.GetDataset()`, then use `spec.NewAbsolutePath()` on the path, then call `Resolve()` on that against `ds.Database()`. Now everything is using the exact same database instance.

Comment (mikegray) [inline cmd/noms/noms_commit.go]: ok, thanks - will do it that way

**Comment by mikegray** [comment]
@aboodman PTAL

**Comment by mikegray** [comment]
`./csv-import --no-progress=1 --commit=0 --dest-type=map:0 sf-registered-businesses/2016-05-15.csv http://localhost:8000::temp | noms commit http://localhost:8000::sf-registered-businesses`

**Comment by mikegray** [comment]
test this

**Comment by mikegray** [comment]
Please test this

**Comment by mikegray** [comment]
build this

**Comment by aboodman** [comment]
@mikegray do you know why the tests are failing?

**Comment by mikegray** [comment]
No, I was unable to get it to build again. I apparently do not know the magic or it is broken. I have another commit coming though anyway - bringing url-fetch, json-import, and xml-import along for the commit ride.

File: cmd/noms/noms_commit.go [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+	allowDupe   bool
+	date        string
+	message     string
+	metaStrings string
+	metaPaths   string
+)
+
+var nomsCommit = &util.Command{
+	Run:       runCommit,
+	UsageLine: "commit [options] <dataset> [absolute-path]",
```
Comment (aboodman) [inline cmd/noms/noms_commit.go]: Can `[absolute-path]` come before `<dataset>`?

File: cmd/noms/noms_commit.go [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+	ds, err := spec.GetDataset(args[0])
+	d.CheckError(err)
+	defer ds.Database().Close()
+
+	var path string
+	if len(args) == 2 {
+		path = args[1]
+	} else {
+		readPath := make([]byte, 256)
+		n, err := os.Stdin.Read(readPath)
```
Comment (aboodman) [inline cmd/noms/noms_commit.go]: It looks like the actual length limit on linux at least is 4k. Maybe use `ReadLine()` instead, or at least make this longer?

Comment (aboodman) [inline cmd/noms/noms_commit.go]: I don't believe that this API is guaranteed to return you a line. You might get less or more than a line. You should definitely use ReadLine().

File: cmd/noms/noms_commit.go [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+	absPath, err := spec.NewAbsolutePath(path)
+	d.CheckError(err)
+
+	value := absPath.Resolve(ds.Database())
+	if value == nil {
+		d.CheckErrorNoUsage(errors.New(fmt.Sprintf("Error resolving value: %s", path)))
+	}
+
+	oldCommitRef, oldCommitExists := ds.MaybeHeadRef()
+	head, ok := ds.MaybeHeadValue()
```
Comment (aboodman) [inline cmd/noms/noms_commit.go]: If the commit exists, there will definitely be a HeadValue. So you don't need both of these.

**Comment by aboodman** [comment]
Really nicely coming along.

**Comment by aboodman** [comment]
BTY

File: cmd/noms/noms_commit.go (hunk 5 above)
Comment (mikegray) [inline cmd/noms/noms_commit.go]: good point, done.

File: cmd/noms/noms_commit.go (hunk 4 above)
Comment (mikegray) [inline cmd/noms/noms_commit.go]: good point, using ReadLine now - done.

File: cmd/noms/noms_commit.go (hunk 3 above)
Comment (mikegray) [inline cmd/noms/noms_commit.go]: sure, makes sense.

**Comment by mikegray** [comment]
@aboodman the build for this PR was broken because master has testExiter removed in favor of the MustRun/Run changes, but this branch was before that. I have rebased to master to pickup those changes which also allows me to turn on some of the disabled tests where stderr was not completely read when failure tests were run. I know this creates more work for you, the reviewer in this case, but having a working build and more fully tested code is better to have for review and landing purposes. PTAL.

**Comment by mikegray** [comment]
@aboodman PTAL - ping

**Comment by aboodman** [comment]
@mikegray I need to run to a meeting. If you have fixed my last comments, LGTM. I will review async.

**Comment by mikegray** [comment]
TBR

**Pull request merged**